### PR TITLE
Benchmark automation

### LINKED ISSUE #899: Benchmark infrastructure
See #898 for background.

We want to be able to run benchmarks as easily as possible, so it's easier for developers to keep track of the performance variance caused by their code.  That entails some form of automation: ideally, it will just be a button click or a script to run.

We also want the benchmarks to run in a realistic environment, i.e. GCE with GCS buckets, or AWS with S3 buckets.  That means that we don't want to naively use CI (which just runs in a single under-powered VM) to run the benchmarks directly.

### PR DESCRIPTION
**Pull request by derekchiang**
Adding a `bench.py` script that runs benchmark on a cloud provider.

Fixes #899.

### CODE REVIEW

File: bench.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    parser = argparse.ArgumentParser(description='Run a Pachyderm benchmark on a cloud provider.', formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    parser.add_argument('--provider', default='GCE', choices=['GCE', 'AWS'], help='the cloud provider to run the benchmark on.')
+    parser.add_argument('--cluster-name', default='pachyderm-benchmark-{}'.format(str(uuid.uuid4())[:8]), help="the name of the cluster to run the benchmark on; one will be created.")
+    parser.add_argument('--cluster-size', default=4, type=int, help='the number of nodes to run the benchmark on.')
+    parser.add_argument('--bucket-name', default='pachyderm-benchmark-bucket-{}'.format(uuid.uuid4()), help='the GCS/S3 bucket to use with the benchmark; one will be created.')
+    parser.add_argument('--volume-name', default='pachyderm-benchmark-volume-{}'.format(uuid.uuid4()), help='the persistent volume to use with the benchmark; one will be created.')
+    parser.add_argument('--benchmark', default='.', help='a regex expression that specifies the benchmark to run; runs all benchmarks by default.')
+    parser.add_argument('--runs', default=1, type=int, help='how many times the benchmark runs.')
```
Comment (sjezewski) [inline bench.py]: This doesn't seem to be used to repeat the benchmark.

File: bench.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    parser.add_argument('--pachd-image', default="pachyderm/pachd:latest", help='the pachd image to use.')
+    parser.add_argument('--job-shim-image', default="pachyderm/job-shim:latest", help='the job-shim image to use.')
+    parser.add_argument('--pachyderm-compile-image', default="pachyderm/pachyderm-compile:latest", help='the pachyderm-compile image to use.')
+
+    args = parser.parse_args()
+
+    print("running the benchmark with the following arguments:")
+    print(args)
```
Comment (sjezewski) [inline bench.py]: I like we're printing this to stdout. If we want to record results we can just pipe to a file and also have the info about how the benchmark was run.

**Comment by derekchiang** [comment]
I don't think using python for scripting is that much different than using bash for scripting, and we do have quite a number of bash scripts.  Using golang for scripting purposes seems a little strange to me.

**Comment by jdoliner** [comment]
I agree in theory that using python for scripting isn't that different from using bash for scripting. But I also see a lot of value in unification so I think the fact that we have a lot of bash scripts is a much stronger argument for more bash than for introducing python.

Looking at the particulars of your script I think it would be cleaner, both in absolute terms and in terms of fitting in better with our codebase, if it was a combination of go and shell/make.

A lot of the python functions are just shelling out to subprocesses which is easier to do in shell/make.

The regex stuff you're doing to get the manifest you want seems fragile to me. Considering that the file you're modifying with regexes is actually programmatically generated by pachctl I think you'd be better off just extending that programmatic generation to do what you want. It's already pretty close I think and this would have the added benefit that this behavior can now be easily used in other places. For example being able to build an image and push it out into a big cluster seems pretty useful for correctness testing as well as benchmarks.

Lastly I'm a little bit dubious of using the `latest` tag and a pull policy of `Always`, that's screwed me so many times in the past that I think it'd be a lot safer to just use explicit tags and bake them into the manifest.

All this being said, @sjezewski's point about prototyping being nicer in scripts still stands. I agree with him that when it comes time to merge this it'd be better not to have the python script but feel free to stick to whatever lets you get something working fast for now and deal with this later. You'll probably know a lot more about the problem constraints by then.

**Comment by derekchiang** [comment]
The `latest` tag is just a default; you can specify whatever images you want to use.  There doesn't seem to be any downside with using `Always` though?  As in, if you use a versioned tag like `1.2.1`, then `Always` does the same thing as `IfNotPresent`.  If you use something like `latest`, then `Always` does the thing you want which is that it will pull the image again even if there's already a `latest` image locally.

Anyways, sounds like you guys would both rather have this be golang/bash, but don't have a strong objection to merging this in for now so we can work on the actual performance stuff.  I'm certainly down to take a look at this again later once this script gets more users and more feedback.

Can I get a LGTM?

**Comment by sjezewski** [comment]
Usually I'd say at the time of PR it should be in a non-scripting language. But it sounds like you want to run w this to try and use it to run actual benchmarks before finalizing it. I'm ok w that in the spirit of 'more frequent PRs' that we've discussed. I'd just request that we make a new GH issue in 1.3 to remember to get to this.

Other than that, LGTM if @jdoliner is good w the `PullPolicy` stuff

image

**Comment by jdoliner** [comment]
Merging to master seems fine I guess, since this won't block development.
Still I feel pretty strongly that keeping our codebase strictly go and shell/make is pretty important so merging stuff like this does make me a little uneasy. How about if we add an issue in the 1.3 milestone to rewrite this?

Also @sjezewski has a few outstanding comments you should respond to. Particularly the one about the unused command line arg.

**Comment by jdoliner** [comment]
Also the downside with `Always`/`:latest` is that there's a lot of ways it can confuse you.
Especially if someone is updating the Docker repo at the same time as you. Pods getting rescheduled can lead to a cluster with mismatched versions. Also there's a bit of lag between when an image is pushed and when it gets published on Docker Hub. I suspect due to caching. So you may run into situations where you rebuild, redeploy, and wind up with the old version of your code. Very easy to get confused and accidentally think that your performance tweaks did nothing.

All of this is avoided if you use a static tag, or maybe even better a hash.

**Pull request merged**